controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentQueue.java

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.controller.deployment;
+
+import com.google.common.collect.ImmutableList;
+import com.yahoo.config.provision.ApplicationId;
+import com.yahoo.searchlib.rankingexpression.rule.Function;
+import com.yahoo.vespa.curator.Lock;
+import com.yahoo.vespa.hosted.controller.Controller;
+import com.yahoo.vespa.hosted.controller.api.integration.BuildService.BuildJob;
+import com.yahoo.vespa.hosted.controller.application.DeploymentJobs.JobType;
+import com.yahoo.vespa.hosted.controller.persistence.CuratorDb;
+
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Stores a queue for each type of job, and offers jobs from each of these to a periodic
+ * polling mechanism which is responsible for triggering the offered jobs in an external build service.
+ *
+ * @author jvenstad
+ */
+public class DeploymentQueue {
+
+    private final Controller controller;
+    private final CuratorDb curator;
+
+    public DeploymentQueue(Controller controller, CuratorDb curator) {
+        this.controller = controller;
+        this.curator = curator;
+    }
+
+    /** Add the given application to the queue of the given job type -- in front if first, at the back otherwise. */
+    public void addJob(ApplicationId applicationId, JobType jobType, boolean first) {
+        locked(jobType, queue -> {
+            if ( ! queue.contains(applicationId)) {
+                if (first)
+                    queue.addFirst(applicationId);
+                else
+                    queue.addLast(applicationId);
+            }
+        });
+    }
+
+    /** List all jobs currently enqueued. */
+    public List<BuildJob> jobs() {
+        ImmutableList.Builder<BuildJob> builder = ImmutableList.builder();
+        for (JobType jobType : JobType.values())
+            for (ApplicationId id : curator.readJobQueue(jobType))
+                toBuildJob(id, jobType).ifPresent(builder::add);
+
+        return builder.build();
+    }
+
+    /** Remove and return a set of jobs to run. This set will contain only one of each job type for capacity constrained zones. */
+    public List<BuildJob> takeJobsToRun() {
+        ImmutableList.Builder<BuildJob> builder = ImmutableList.builder();
+        for (JobType jobType : JobType.values())
+            locked(jobType, queue ->
+                    queue.stream()
+                            .limit(isCapacityConstrained(jobType) ? 1 : 1 << 30)
```

Comment:
Replace magic value with `Long.MAX_VALUE`. Easier to read as the limit will be "1 or max".

Reply:
Yes, of course.